### PR TITLE
Add the ability to remove a runtime field

### PR DESCRIPTION
**Pull request by javanna**
Currently, existing runtime fields can be updated, but they cannot be removed. That allows to correct potential mistakes, but once a runtime field is added to the index mappings, it is not possible to remove it.

With this commit we introduce the ability to remove an existing runtime field by providing a null value for it through the put mapping API. If a field with such name does not exist, such specific instruction will have no effect on other existing runtime fields.

Note that the removal of runtime fields makes the recently introduced assertRefreshItNotNeeded assertion trip, because when each local node merges mappings back in, the runtime fields that were previously removed by the master node, get added back again locally. This is only a problem for the assertion that verifies that the removed refresh operation is never needed. We worked around this by tweaking the assertion to ignore runtime fields completely, for simplicity, by assertion on the serialized merged mappings and incoming mappings without the corresponding runtime section.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -202,13 +202,20 @@ public void updateMapping(final IndexMetadata currentIndexMetadata, final IndexM
...
         CompressedXContent mergedMappingSource;
         try {
-            mergedMappingSource = new CompressedXContent(mergedMapping, XContentType.JSON, ToXContent.EMPTY_PARAMS);
+            mergedMappingSource = new CompressedXContent(mergedMapping, XContentType.JSON, params);
+        } catch (Exception e) {
+            throw new AssertionError("failed to serialize source for type [" + type + "]", e);
+        }
+        CompressedXContent incomingMappingSource;
+        try {
+            incomingMappingSource = new CompressedXContent(incomingMapping, XContentType.JSON, params);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:218]: There were a couple of ways to work around this. The assertion was recently added (as the refresh mapping action was just removed) hence it is important to keep it around for some time. It seems rather safe to exclude runtime fields from the check. As for how to exclude them, we could have parsed them back into a map and filter its content, or recreate Mapping with a modified root object mapper that has empty runtime fields. Both of these approaches ended up being more complex that I initially considered. 

This approach has the risk of introducing a custom toxcontent param, which must never be used outside of this context, but it does its job and is relatively simple. I am curious to hear other's thoughts on this. @DaveCTurner @ywelsch too  ;)

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java [hunk 2]
```diff
@@ -326,7 +326,13 @@ protected void doMerge(ObjectMapper mergeWith, MergeReason reason) {
             }
         }
         assert this.runtimeFieldTypes != mergeWithObject.runtimeFieldTypes;
-        this.runtimeFieldTypes.putAll(mergeWithObject.runtimeFieldTypes);
+        for (Map.Entry<String, RuntimeFieldType> runtimeField : mergeWithObject.runtimeFieldTypes.entrySet()) {
+            if (runtimeField.getValue().isPlaceholderForRemoval()) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:330]: Possibly dumb question: is there any reason we can't just use `null` as the placeholder here?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 3]
```diff
@@ -202,13 +202,20 @@ public void updateMapping(final IndexMetadata currentIndexMetadata, final IndexM
...
 
     private boolean assertRefreshIsNotNeeded(DocumentMapper currentMapper,
-                                          String type,
-                                          CompressedXContent incomingMappingSource,
-                                          Mapping incomingMapping) {
+                                             String type,
+                                             Mapping incomingMapping) {
         Mapping mergedMapping = mergeMappings(currentMapper, incomingMapping, MergeReason.MAPPING_RECOVERY);
+
+        ToXContent.MapParams params = new ToXContent.MapParams(Collections.singletonMap("skip_runtime_section", "true"));
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:209]: Nice! Might be worth making `skip_runtime_section` a static final String on RuntimeFieldType and adding a comment there on why we need it?

File: server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java [hunk 4]
```diff
@@ -669,9 +683,9 @@ public void testRuntimeSectionMissingType() throws IOException {
     }
 
     public void testRuntimeSectionWrongFormat() throws IOException {
-        XContentBuilder mapping = runtimeMapping(builder -> builder.field("field", "value"));
+        XContentBuilder mapping = runtimeMapping(builder -> builder.field("field", 123));
         MapperParsingException e = expectThrows(MapperParsingException.class, () -> createMapperService(mapping));
-        assertEquals("Failed to parse mapping: Expected map for runtime field [field] definition but got a java.lang.String",
+        assertEquals("Failed to parse mapping: Expected map for runtime field [field] definition but got a java.lang.Integer",
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:688]: Is this related to the rest of the PR?

**Review by romseygeek — COMMENTED** [review]
I like the general direction! I left a couple of questions.

File: server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java (hunk 4 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:688]: good question: a little, or maybe not. We used to error when null was provided (same error as when something that is not a map is provided), but I figured the error message contained the type of the key (always string) rather than that of the value :) We even had a test for it which was working because the value was also a string, so I changed it to integer just to make sure.

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java (hunk 2 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:330]: This is all but a dumb question! This used to be necessary because we'd build a DocumentMapper, which includes not only Mapping but also MappingLookup and FieldTypeLookup, and they could not really support a null type. At this point though, I may have made that unnecessary without realizing. Let me give this a try.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:330]: Unfortunately the placeholder is still needed. Once we move away from constructing DocumentMapper before merging, and we will remove DocumentMapper#merge, we will be able to easily remove the placeholder. I will note this down.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:340]: actually, I looked even more closely, and it seems pretty easy to address this, I will open a separate PR for it.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java]: I have removed the placeholder, it is no longer needed after merging #69005 in.

**Review by romseygeek — APPROVED** [review]
LGTM!

File: docs/reference/mapping/runtime.asciidoc [hunk 5]
```diff
@@ -139,18 +139,34 @@ PUT my-index
 }
 ----
 
+Runtime fields can be updated at any point in time: whenever a runtime field
+is added with the same name as an already existing field, the new field will
+overwrite the existing one.
+Existing runtime fields can also be removed from the index mappings by setting
+them to null as follows:
```
Comment (lockewritesdocs) [inline docs/reference/mapping/runtime.asciidoc:146]:
```suggestion
You can update or remove runtime fields at any time. To replace an existing runtime field, add a new runtime field to the mappings with the same name. To remove a runtime field from the mappings, set the value of the runtime field to `null`:
```
